planner_cspace/src/planner_3d.cpp

```diff
@@ -394,6 +412,128 @@ class Planner3dNode
+    e[2] = 0;
+
+    if (!(cm_rough_.validate(s) && cm_rough_.validate(e)))
+    {
+      ROS_ERROR("Given start or goal is not on the map.");
+      return false;
+    }
+    else if (cm_rough_[s] == 100 || cm_rough_[e] == 100)
+    {
+      ROS_ERROR("Given start or goal is in Rock.");
+      return false;
+    }
+
+    const Astar::Vecf euclid_cost_coef = ec_rough_;
+
+    const auto cbCost = [this, &euclid_cost_coef](
+        const Astar::Vec &s, Astar::Vec &e,
+        const Astar::Vec &v_goal, const Astar::Vec &v_start,
+        const bool hyst) -> float
+    {
+      const Astar::Vec d = e - s;
+      float cost = euclidCost(d, euclid_cost_coef);
+
+      int sum = 0;
+      int num = 0;
+      const auto cache_page = linear_interp_cache_[0].find(d);
+      if (cache_page == linear_interp_cache_[0].end())
+        return -1;
+      for (const auto &pos_diff : cache_page->second)
+      {
+        const int posi[3] =
+            {
+              s[0] + pos_diff[0], s[1] + pos_diff[1], 0
+            };
+        const Astar::Vec pos(posi);
+        const auto c = cm_rough_[pos];
+        if (c > 99)
+          return -1;
+        sum += c;
+        num++;
+      }
+      const float &distf = linear_distf_cache_[0][d];
+      cost += sum * map_info_.linear_resolution * distf * cc_.weight_costmap_ / (100.0 * num);
+
+      return cost;
+    };
+    const auto cbCostEstim = [](const Astar::Vec &s, const Astar::Vec &e)
+    {
+      auto s2 = s;
+      s2[2] = 0;
+      auto cost = sqrtf((e - s2).sqlen());
+
+      if (cost == FLT_MAX)
+        return FLT_MAX;
+      return cost;
+    };
+    const auto cbSearch = [this](
+        const Astar::Vec &p,
+        const Astar::Vec &s, const Astar::Vec &e) -> std::vector<Astar::Vec>&
+    {
+      return search_list_rough_;
+    };
+    const auto cbProgress = [](const std::list<Astar::Vec> &path_grid)
+    {
+      return true;
+    };
+
+    const auto ts = boost::chrono::high_resolution_clock::now();
+    // ROS_INFO("Planning from (%d, %d, %d) to (%d, %d, %d)",
+    //   s[0], s[1], s[2], e[0], e[1], e[2]);
+    std::list<Astar::Vec> path_grid;
+    if (!as_.search(s, e, path_grid,
+                    std::bind(cbCost,
+                              std::placeholders::_1, std::placeholders::_2,
+                              std::placeholders::_3, std::placeholders::_4, false),
+                    std::bind(cbCostEstim,
+                              std::placeholders::_1, std::placeholders::_2),
+                    std::bind(cbSearch,
+                              std::placeholders::_1,
+                              std::placeholders::_2, std::placeholders::_3),
+                    std::bind(cbProgress,
+                              std::placeholders::_1),
+                    0,
+                    1.0f / freq_min_,
+                    find_best_))
+    {
+      ROS_WARN("Path plan failed (goal unreachable)");
+      if (!find_best_)
```

Comment:
This aims to update costmap by motion if the goal is not reachable.
In this service, it would be better always to return false.